### PR TITLE
Core implementation of GossipProtocol

### PR DESCRIPTION
**Pull request by kpudlik**

### CODE REVIEW

File: src/main/scala/org/constellation/gossip/GossipPathState.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.constellation.gossip
+
+object GossipPathState extends Enumeration {
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/gossip/GossipPathState.scala:3]: Please use enumeratum (see NodeState)

File: src/main/scala/org/constellation/gossip/GossipService.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import cats.effect.Concurrent
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import org.constellation.gossip.sampling.PeerSampling
+import org.constellation.infrastructure.p2p.ClientInterpreter
+import org.constellation.schema.Id
+
+abstract class GossipService[F[_], A](selfId: Id, peerSampling: PeerSampling[F, Id])(implicit val F: Concurrent[F]) {
+
+  private val clients: Ref[F, Map[Id, ClientInterpreter[F]]] = Ref.unsafe(Map.empty[Id, ClientInterpreter[F]])
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/gossip/GossipService.scala:12]: Why do you need to store same ClientInterpreter for every Id? ClientInterpreter is a singleton afaik. 
IMO you should store here peer metadata (host, port, id) which will be applied to a proper api call (as providing a context into a run method)

**Review by marcinwadon — CHANGES_REQUESTED** [review]

File: src/test/scala/org/constellation/gossip/RandomPeerSamplingTest.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      duplicates.size shouldBe 0
+    }
+
+    "start and ends with selfId" in {
+      val nodes = Set(Id("b"), Id("c"), Id("d"), Id("e"), Id("f"), Id("g"), Id("h"))
+      val fanout = 2
+      peerSampling.updatePeerList(nodes).unsafeRunSync()
+      val path = peerSampling.selectPaths(fanout).unsafeRunSync()
+
+      println(path)
```
Comment (tbekas) [inline src/test/scala/org/constellation/gossip/RandomPeerSamplingTest.scala:65]: println

**Review by tbekas — CHANGES_REQUESTED** [review]

File: src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  private val peerList: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+
+  override def updatePeerList(peers: Set[Id]): F[Unit] =
+    peerList.modify { _ =>
+      (peers, ())
+    }
+
+  def selectPaths(fanout: Int): F[Seq[List[Id]]] =
+    peerList.modify { p =>
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala:23]: Modify uses loop and does not get data if any change is in progress. Get takes data immediately. There is an explanation of it in the Ref scaladoc, I will find it later.

File: src/main/scala/org/constellation/gossip/GossipPath.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  def isPrev(id: Id): Boolean = prev.contains(id)
+
+  def prev: Option[Id] = path.lift(cursor - 1)
+}
+
+object GossipPath {
+
+  @tailrec
+  def apply(path: List[Id]): GossipPath = GossipPath(path)
```
[the remark this answers is not in the document]
Comment (kpudlik) [reply src/main/scala/org/constellation/gossip/GossipPath.scala:33]: Yep, I forgot about `new` so Intellij autoincluded tailrec

File: src/main/scala/org/constellation/gossip/GossipService.scala (hunk 2 above)
Comment (kpudlik) [reply src/main/scala/org/constellation/gossip/GossipService.scala:12]: I mislooked ClientInterpreter with PeerClientMetadata :D

File: src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (tbekas) [reply src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala:23]: It makes no difference, except for the performance penalty on read. If there's concurrent write happening just after the read has been done, you're going to process stale values anyway. Locking is required when modifying the state, hence this method is named `modify`.

[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala]: up to you :)

**Review by marcinwadon — APPROVED** [review]

**Review by tbekas — DISMISSED** [review]

File: src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    partition(Random.shuffle(peers.toIndexedSeq), fanout)
+      .map(selfId +: _ :+ selfId)
+      .map(IndexedSeq[Id](_: _*))
+  }
+
+  private def getPeers: F[Set[Id]] =
+    cluster.getPeerInfo
+      .map(_.filter { case (_, pd) => NodeState.isNotOffline(pd.peerMetadata.nodeState) }.keySet)
```
Comment (TheMMaciek) [inline src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala:35]: Maybe with the push strategy we should not be sending data to a leaving peer?

**Review by TheMMaciek — DISMISSED** [review]

File: src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala (hunk 6 above)
Comment (kpudlik) [reply src/main/scala/org/constellation/gossip/sampling/RandomPeerSampling.scala:35]: fetching peer proposals uses only `isNotOffline` check and it seems to be working 🤔 but you may be right. We need to test that.

**Review by marcinwadon — APPROVED** [review]

**Pull request merged**